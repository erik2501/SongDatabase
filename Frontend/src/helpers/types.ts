
export type Song ={
    songID: number,
    songName: string,
    artistName: string
}