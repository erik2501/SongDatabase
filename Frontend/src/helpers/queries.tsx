import { gql } from '@apollo/client';


export const GET_REVIEWS = gql`
    query GetReviews ($songID: Int!, $amount: Int!) {
        reviewsBySongID (songID: $songID, amount: $amount) {
            userName
            star
            description
            songID
        }
    }
    `

export const GET_COUNT = gql`
    query ( $searchWord: String, $year: Int ){
        songSearchCount( searchWord: $searchWord, year: $year )
    }
`;

export const CREATE_REVIEW = gql`
    mutation addReviewForSong($userName: String!, $star: Int!, $description: String!, $songID: Int!) {
        addReview(userName: $userName, star: $star, description: $description, songID: $songID) {
            userName
            star
            description
            songID
        }
    }`

export const GET_AVG_REVIEW_SCORE = gql`
    query ( $songID: Int ){
        reviewAvgScoreBySongID( songID:$songID ) {
        avgScore
        }
    }
`;

export const GET_SONG_BY_SONGID = gql`
    query GetSong ($songID: Int!) {
        songBySongID (songID: $songID) {
            songID
            songName
            artistName
            durationMS
            year
            energy
            imageURL
        }
    }
`;

export const GET_SEARCH = gql`
    query Get_Search ($searchWord: String, $skip: Int, $amount:Int, $year: Int, $order: Int){
        songSearch(skip: $skip, amount:$amount, searchWord: $searchWord, year: $year, order: $order ) {
            songID
            artistName
            songName
            imageURL
        }
    }
`;

export const GET_DISTINCT_YEARS = gql`
    query {
        getDistinctYears
    }
`;