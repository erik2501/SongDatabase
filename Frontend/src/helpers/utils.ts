// Her har vi en debounce funksjon som gjør at rendering av den gjeldende kompnenten venter 500 millisekunder. 
// I hovedsak bruker vi dette til å vente med å hente sanger 500 ms etter bruker har begynt å skrive noe i søk. 

export function debounce<T extends (...argsx: any[]) => void>(callback: T, delay = 500) {
    let timeout: NodeJS.Timeout;
  
    return (...args: Parameters<T>) => {
      clearTimeout(timeout);
      timeout = setTimeout(() => {
        callback(...args);
      }, delay);
    };
  }