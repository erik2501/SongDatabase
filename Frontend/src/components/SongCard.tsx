import { Song } from '../helpers/types';
import { useNavigate } from 'react-router-dom';

const SongCard = ({song}: {song: Song}) => {


    const nav = useNavigate();
    const updateNav = () => {
        nav('/song/' + song.songID);
    };


    return(
        <div>
            <button className='songCard' onClick={() => updateNav()}>
                <p>{'Song: ' + song.songName}</p>
                <p>{'Artist: ' + song.artistName}</p>
            </button>
        </div>
        
    )
}

export default SongCard;