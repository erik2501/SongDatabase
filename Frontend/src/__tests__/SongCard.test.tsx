import SongCard from "../components/SongCard"
import renderer from 'react-test-renderer'
import { Song } from '../helpers/types'
import { MockedProvider } from "@apollo/client/testing";
import { GET_AVG_REVIEW_SCORE } from "../helpers/queries";
import { BrowserRouter } from "react-router-dom";

const dummySong: Song = {
    songID: 0,
    songName: "Oops!...I Did It Again",
    artistName: "Britney Spears",
    imageURL: "https://upload.wikimedia.org/wikipedia/en/2/24/Britney_Spears_-_Oops%2"
}

// reviewAvgScoreBySongID
const mocks = [{
    request: {
        query: GET_AVG_REVIEW_SCORE,
        variables: {
            songID: dummySong.songID
        }
    },
    result: {
        data: {
            reviewAvgScoreBySongID: [
                {
                    __typename: 'avgscore',
                    avgScore: 3.5
                }

            ]
        }
    }
}];

it('renders when passed a song', () => {
    const tree = renderer.create((<BrowserRouter><MockedProvider mocks={mocks} addTypename={false}><SongCard song={dummySong} /></MockedProvider></BrowserRouter>)).toJSON();
    expect(tree).toMatchSnapshot();
});
