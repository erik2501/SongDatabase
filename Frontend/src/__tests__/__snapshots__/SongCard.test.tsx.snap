// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when passed a song 1`] = `
<a
  className="songCardContainer"
  onClick={[Function]}
  onKeyUp={[Function]}
>
  <div
    className="songCard"
    tabIndex={0}
  >
    <img
      className="albumcoverImage"
      src="https://upload.wikimedia.org/wikipedia/en/2/24/Britney_Spears_-_Oops%2"
    />
    <div
      className="textbox"
    >
      <h2>
        Oops!...I Did It Again
      </h2>
      <p>
        Artist: Britney Spears
      </p>
    </div>
    <div
      className="avgScoreContainer"
    />
  </div>
</a>
`;

exports[`renders when passed name and rating and description 1`] = `
<div>
  <div
    className="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    style={
      Object {
        "paddingBottom": "15px",
      }
    }
  >
    <h5>
      snapshotTestName
    </h5>
    <span
      aria-label="1 Star"
      className="MuiRating-root MuiRating-sizeMedium MuiRating-readyOnly css-ryrseu-MuiRating-root"
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      role="img"
    >
      <span>
        <span
          className="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
      </span>
      <span>
        <span
          className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
      </span>
      <span>
        <span
          className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
      </span>
      <span>
        <span
          className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
      </span>
      <span>
        <span
          className="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="StarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
        </span>
      </span>
    </span>
    <p>
      bra
    </p>
  </div>
</div>
`;
